build.sh

```diff
@@ -2,53 +2,200 @@
+            # Install silently, redirecting output to /dev/null, but keep errors
+            if sudo pacman -S --noconfirm "$package" &> /dev/null; then
+                echo_plain "Package $package installed successfully."
+            else
+                echo_error "Failed to install $package."
+            fi
+        else
+            echo_plain "$package is already installed."
+        fi
+    done
+}
+
+
+# Function to install dependencies for most Linux systems
+install_dependencies_for_most_linux_systems() {
+  echo_success "Checking dependencies"
+
   if command -v apk &> /dev/null; then
-      echo "Detected apk"
-      sudo apk add python3 py3-pip py3-virtualenv
+      install_via_apk
   elif command -v apt &> /dev/null; then
-      echo "Detected apt"
-      sudo apt install python3-pip python3-virtualenv
+      install_via_apt
   elif command -v pacman &> /dev/null; then
-      echo "Detected pacman"
-      sudo pacman -S --noconfirm python-pip python-virtualenv
+      install_via_pacman
   else
-      echo "Unsupported package manager. Please install the packages and create the virtualenv manually."
+      echo_error "Unsupported package manager. Please install the packages and create the virtualenv manually."
       exit 1
   fi
-  echo -e "${YELLOW}Creating and activating a virtual build environment in \"$UBPORTSDOCSENV\".${PLAIN}"
-  virtualenv $UBPORTSDOCSENV
-  source $UBPORTSDOCSENV/bin/activate
-  echo -e "${YELLOW}Installing build tools and prerequisites (using pip).${PLAIN}"
-  pip3 install -r requirements.txt
-fi
-
-echo -e "${GREEN}Building...${PLAIN}"
-rm -rf _build/
-if [ "$(uname)" == "Linux" ]; then
-  sphinx-build -Wa . _build/html -j `nproc --all` || exit $?
-else
-  sphinx-build -Wa . _build/html -j `sysctl -n hw.ncpu` || exit $?
-fi
-sphinx-build -Wab rediraffecheckdiff . _build/html
-
-echo "Open _build/html/index.html in your browser to navigate through the docs"
+}
+#endregion INSTALL_DEPENDENCIES
+
+#region PREPARE_VIRTUALENV
+prepare_virtualenv_for_python(){
+  # check if virtualenv already exists
+  if [ -d "$UBPORTSDOCSENV" ]; then
+    echo_success "Activating virtual build environment found in \"$UBPORTSDOCSENV\"."
+
+    source $UBPORTSDOCSENV/bin/activate
+  else
+    echo_warning "Build environment not found in \"$UBPORTSDOCSENV\" ... creating it."
+    echo_warning "Creating and activating a virtual build environment in \"$UBPORTSDOCSENV\"."
+
+    virtualenv $UBPORTSDOCSENV
+    source $UBPORTSDOCSENV/bin/activate
+
+    echo_warning "Installing build tools and prerequisites (using pip)."
+
+    pip3 install -r requirements.txt
+  fi
+}
+#endregion PREPARE_VIRTUALENV
+
+#region BUILD
+wrapper_build_documentation(){
+  rm -rf _build/
+  if [ "$(uname)" == "Linux" ]; then
+    sphinx-build -Wa . _build/html -j `nproc --all` || exit $?
+  else
+    sphinx-build -Wa . _build/html -j `sysctl -n hw.ncpu` || exit $?
+  fi
+  sphinx-build -Wab rediraffecheckdiff . _build/html
```

Comment:
Should this also `|| exit` ? 

That should also allow you to put a echo_success after